sstable/reader_virtual.go

```diff
@@ -204,10 +248,264 @@ func (v *virtualState) constrainBounds(
+		f = append(append([]byte{}, v.vState.prefixChange.Backing...), f[len(v.vState.prefixChange.Materialized):]...)
+		l = append(append([]byte{}, v.vState.prefixChange.Backing...), l[len(v.vState.prefixChange.Materialized):]...)
+	}
+
 	return v.reader.EstimateDiskUsage(f, l)
 }
 
 // CommonProperties implements the CommonReader interface.
 func (v *VirtualReader) CommonProperties() *CommonProperties {
 	return &v.Properties
 }
+
+type prefixReplacingIterator struct {
+	i         Iterator
+	cmp       base.Compare
+	src, dst  []byte
+	arg, arg2 []byte
+	res       InternalKey
+	err       error
+}
+
+var errInputPrefixMismatch = errors.New("key argument does not have prefix required for replacement")
+var errOutputPrefixMismatch = errors.New("key returned does not have prefix required for replacement")
+
+var _ Iterator = (*prefixReplacingIterator)(nil)
+
+// NewPrefixReplacingIterator wraps an iterator over keys that have prefix `src`
+// in an iterator that will make them appear to have prefix `dst`. Every key
+// passed as an argument to methods on this iterator must have prefix `dst`, and
+// every key produced by the underlying iterator must have prefix `src`.
+func NewPrefixReplacingIterator(i Iterator, src, dst []byte, cmp base.Compare) Iterator {
+	return &prefixReplacingIterator{
+		i:   i,
+		cmp: cmp,
+		src: src, dst: dst,
+		arg: append([]byte{}, src...), arg2: append([]byte{}, src...),
+		res: InternalKey{UserKey: append([]byte{}, dst...)},
+	}
+}
+
+func (p *prefixReplacingIterator) SetContext(ctx context.Context) {
+	p.i.SetContext(ctx)
+}
+
+func (p *prefixReplacingIterator) rewriteArg(key []byte) []byte {
+	if !bytes.HasPrefix(key, p.dst) {
+		p.err = errInputPrefixMismatch
+		return key
+	}
+	p.arg = append(p.arg[:len(p.src)], key[len(p.dst):]...)
+	return p.arg
+}
+
+func (p *prefixReplacingIterator) rewriteArg2(key []byte) []byte {
+	if !bytes.HasPrefix(key, p.dst) {
+		p.err = errInputPrefixMismatch
+		return key
+	}
+	p.arg2 = append(p.arg2[:len(p.src)], key[len(p.dst):]...)
+	return p.arg2
+}
+
+func (p *prefixReplacingIterator) rewriteResult(
+	k *InternalKey, v base.LazyValue,
+) (*InternalKey, base.LazyValue) {
+	if k == nil {
+		return k, v
+	}
+	if !bytes.HasPrefix(k.UserKey, p.src) {
+		panic(errOutputPrefixMismatch)
+	}
+	p.res.Trailer = k.Trailer
+	p.res.UserKey = append(p.res.UserKey[:len(p.dst)], k.UserKey[len(p.src):]...)
+	return &p.res, v
+}
+
+// SeekGE implements the Iterator interface.
+func (p *prefixReplacingIterator) SeekGE(
+	key []byte, flags base.SeekGEFlags,
+) (*InternalKey, base.LazyValue) {
+	cmp := p.cmp(key, p.dst)
+	if cmp < 0 {
+		return p.rewriteResult(p.i.SeekGE(p.src, flags))
+	}
+	return p.rewriteResult(p.i.SeekGE(p.rewriteArg(key), flags))
+}
+
+// SeekPrefixGE implements the Iterator interface.
+func (p *prefixReplacingIterator) SeekPrefixGE(
+	prefix, key []byte, flags base.SeekGEFlags,
+) (*InternalKey, base.LazyValue) {
+	cmp := p.cmp(key, p.dst)
+	if cmp < 0 {
```

Comment:
I think the issue here is that the comparator isn't necessarily defined on `p.dst`; we require it to be defined on the prefix as returned by `Comparer.Split`, but this prefix could be much shorter than that. I would remove this optimization and just call the child method always.